From 1d94d045b0884dd0722aa0e9308409fe0312d891 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?S=C3=B6ren=20Tempel?= <soeren+git@soeren-tempel.net>
Date: Sun, 27 Apr 2025 16:40:25 +0200
Subject: [PATCH] librc: dirfd is initalized with zero not with -1

Without this patch, this code therefore closes FD 0 repeatedly.

See https://github.com/OpenRC/openrc/pull/851

This fixes a regression from 27eae2c380239e7baf113c1b228c6fe679e6a541.
---
 src/librc/librc.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/src/librc/librc.c b/src/librc/librc.c
index 879e0171..0ea21d56 100644
--- a/src/librc/librc.c
+++ b/src/librc/librc.c
@@ -638,8 +638,11 @@ rc_set_user(void)
 	rc_dirs.scriptdirs[SCRIPTDIR_SVC] = rc_dirs.svcdir;
 
 	for (size_t i = 0; i < RC_DIR_MAX; i++) {
-		if (dirfds[i] == -1)
+		/* dirfd is zero-initialized, though zero is a valid
+		 * FD, dirfd should never contain FDs which are <= 2. */
+		if (dirfds[i] <= 0)
 			continue;
+
 		close(dirfds[i]);
 		dirfds[i] = -1;
 	}
